clang/lib/CIR/CodeGen/CIRGenTBAA.h

```diff
@@ -13,16 +13,231 @@
+
+  TBAAAccessInfo()
+      : TBAAAccessInfo(/* accessType= */ nullptr, /* size= */ 0) {};
+
+  static TBAAAccessInfo getMayAliasInfo() {
+    return TBAAAccessInfo(TBAAAccessKind::MayAlias, /* baseType= */ {},
+                          /* accessType= */ nullptr,
+                          /* offset= */ 0, /* size= */ 0);
+  }
+
+  bool isMayAlias() const { return kind == TBAAAccessKind::MayAlias; }
+
+  static TBAAAccessInfo getIncompleteInfo() {
+    return TBAAAccessInfo(TBAAAccessKind::Incomplete, /* baseType= */ {},
+                          /* accessType= */ {},
+                          /* offset= */ 0, /* size= */ 0);
+  }
+
+  bool isIncomplete() const { return kind == TBAAAccessKind::Incomplete; }
+
+  bool operator==(const TBAAAccessInfo &other) const {
+    return kind == other.kind && baseType == other.baseType &&
+           accessType == other.accessType && offset == other.offset &&
+           size == other.size;
+  }
+
+  bool operator!=(const TBAAAccessInfo &other) const {
+    return !(*this == other);
+  }
+
+  explicit operator bool() const { return *this != TBAAAccessInfo(); }
+
+  /// kind - The kind of the access descriptor.
+  TBAAAccessKind kind;
+
+  /// baseType - The base/leading access type. May be null if this access
+  /// descriptor represents an access that is not considered to be an access
+  /// to an aggregate or union member.
+  mlir::LLVM::TBAATypeDescriptorAttr baseType;
+
+  /// accessType - The final access type. May be null if there is no TBAA
+  /// information available about this access.
+  mlir::LLVM::TBAATypeDescriptorAttr accessType;
+
+  /// offset - The byte offset of the final access within the base one. Must be
+  /// zero if the base access type is not specified.
+  uint64_t offset;
+
+  /// size - The size of access, in bytes.
+  uint64_t size;
+};
 
 /// CIRGenTBAA - This class organizes the cross-module state that is used while
 /// lowering AST types to LLVM types.
-class CIRGenTBAA {};
+class CIRGenTBAA {
+  mlir::MLIRContext *ctx;
+  clang::ASTContext &Context;
+  CIRGenTypes &CGTypes;
+  mlir::ModuleOp Module;
+  const clang::CodeGenOptions &CodeGenOpts;
+  const clang::LangOptions &Features;
 
-} // namespace cir
+  /// MetadataCache - This maps clang::Types to scalar TBAATypeDescriptorAttr
+  /// describing them.
+  llvm::DenseMap<const clang::Type *, mlir::LLVM::TBAATypeDescriptorAttr>
+      MetadataCache;
+  /// This maps clang::Types to a base access type in the type DAG.
+  llvm::DenseMap<const clang::Type *, mlir::LLVM::TBAATypeDescriptorAttr>
+      BaseTypeMetadataCache;
+  /// This maps TBAA access descriptors to tag nodes.
+  llvm::DenseMap<TBAAAccessInfo, mlir::ArrayAttr> AccessTagMetadataCache;
+
+  /// StructMetadataCache - This maps clang::Types to TBAATypeDescriptorAttr
+  /// describing them for struct assignments.
+  llvm::DenseMap<const clang::Type *, mlir::ArrayAttr> StructMetadataCache;
 
+  mlir::LLVM::TBAARootAttr root;
```

Comment:
Please remove all member repetions in the comments, i.e. `StructMetadataCache - ` and all others. LLVM deprecated this already and we don't want to generate extra burden when upstreaming. This should apply everywhere else in this PR, not only this file.